base/compiler/tfuncs.jl

```diff
@@ -484,14 +484,27 @@ function typeof_tfunc(@nospecialize(t))
 end
 add_tfunc(typeof, 1, 1, typeof_tfunc, 0)
 
-function typeassert_tfunc(@nospecialize(v), @nospecialize(t))
-    t = instanceof_tfunc(t)[1]
-    t === Any && return v
+function typeassert_instance(@nospecialize(v), @nospecialize(t))
     if isa(v, Const)
         if !has_free_typevars(t) && !isa(v.val, t)
             return Bottom
         end
         return v
+    elseif isa(v, PartialStruct)
+        has_free_typevars(t) && return v
+        widev = widenconst(v)
+        newT = typeintersect(widev, t)
+        if newT === widev
+            return v
+        end
+        new_fields = Vector{Any}(undef, length(v.fields))
+        for i = 1:length(new_fields)
+            new_fields[i] = typeassert_instance(v.fields[i], getfield_tfunc(newT, Const(i)))
+            if new_fields[i] === Bottom
+                return Bottom
+            end
+        end
+        return PartialStruct(newT, new_fields)
```

Comment:
```suggestion
        return tuple_tfunc(new_fields)
```

This would be strictly more correct.

Reply:
`t` is not necessarily a tuple